File: ui/TodoPanel.lua
-- ui/TodoPanel.lua
-- The "To Do" tab: everything worth doing about your gear, in the order that unblocks it.
--
-- Valuate:BuildTodoList has answered this question since it was written, and the answer only
-- ever reached a chat frame. Worse, the login line announced the list and then asked you to
-- go and type a command to read it:
--
--     [Valuate] 3 things worth doing - /valuate todo
--
-- Every item it produces already carries a `command` - it was built to be clicked from the
-- start. This is that.
--
-- The panel holds NO opinions about priority. Order, wording and what counts as worth doing
-- are all decided in BuildTodoList, which explains its own reasoning at length and is tested
-- by tools/todotest.js. Duplicating any of that here would give the addon two answers to the
-- same question, and the one on screen would be the one nobody was testing.

local _, ns = ...

local PADDING, ELEMENT_SPACING = ns.PADDING, ns.ELEMENT_SPACING
local COLORS = ns.COLORS
local BACKDROP_PANEL = ns.BACKDROP_PANEL
local FONT_H2, FONT_BODY, FONT_SMALL = ns.FONT_H2, ns.FONT_BODY, ns.FONT_SMALL

-- A FLOOR, not the height. Rows grow to fit their own text; see Refresh.
local ROW_HEIGHT = 40
local ROW_GAP = 6
local TEXT_TOP_PAD = 8
local TEXT_BOTTOM_PAD = 8
local DETAIL_GAP = 3

-- Rows are POOLED, not created per refresh.
--
-- The list is rebuilt every time the tab is opened, because a to-do list that is stale is
-- worse than none - it would have you chasing an upgrade you already equipped. Creating
-- frames on each rebuild would leak one row per visit for the whole session; WoW frames
-- cannot be destroyed, only hidden and reused.
local rowPool = {}

local function BuildRow(parent, index)
    local row = CreateFrame("Frame", nil, parent)
    -- Marked so a gate can find these among the window's several hundred frames. The rows
    -- are anonymous and pooled, and identifying them by size or parent would break the
    -- moment either changed for a layout reason.
    row.__todoRow = true
    row:SetHeight(ROW_HEIGHT)
    row:SetPoint("LEFT", parent, "LEFT", 0, 0)
    row:SetPoint("RIGHT", parent, "RIGHT", 0, 0)
    if index == 1 then
        row:SetPoint("TOP", parent, "TOP", 0, 0)
    else
        row:SetPoint("TOP", rowPool[index - 1], "BOTTOM", 0, -ROW_GAP)
    end
    row:SetBackdrop(BACKDROP_PANEL)
    row:SetBackdropColor(unpack(COLORS.panelBg))
    row:SetBackdropBorderColor(unpack(COLORS.borderDark))

    -- The accent says WHICH kind of thing this is at a glance, and is coloured per kind by
    -- Refresh. A stale scale and a free upgrade want to look different before you read them.
    local accent = row:CreateTexture(nil, "OVERLAY")
    accent:SetPoint("TOPLEFT", row, "TOPLEFT", 0, 0)
    accent:SetPoint("BOTTOMLEFT", row, "BOTTOMLEFT", 0, 0)
    accent:SetWidth(3)
    row.accent = accent

    local text = row:CreateFontString(nil, "OVERLAY", FONT_BODY)
    text:SetPoint("TOPLEFT", row, "TOPLEFT", 12, -TEXT_TOP_PAD)
    text:SetPoint("RIGHT", row, "RIGHT", -120, 0)
    text:SetJustifyH("LEFT")
    text:SetTextColor(unpack(COLORS.textTitle))
    row.text = text

    local detail = row:CreateFontString(nil, "OVERLAY", FONT_SMALL)
    detail:SetPoint("TOPLEFT", text, "BOTTOMLEFT", 0, -DETAIL_GAP)
    detail:SetPoint("RIGHT", row, "RIGHT", -120, 0)
    detail:SetJustifyH("LEFT")
    detail:SetTextColor(unpack(COLORS.textDim))
    row.detail = detail

    -- One button per row, running that item's own command. The command comes from the item,
    -- not from a table here: BuildTodoList decides what each entry means, and a second
    -- mapping in this file would be free to disagree with it.
    local go = ns.CreateStyledButton(row, "Show me", 96, 20)
    go:SetPoint("RIGHT", row, "RIGHT", -10, 0)
    row.go = go

    -- Handlers installed ONCE, reading row.command, which Refresh updates.
    --
    -- Rows are pooled, so anything installed inside Refresh is installed again on every
    -- visit to the tab. For SetScript that merely replaces; for HookScript it CHAINS, and
    -- a row would accumulate one more tooltip handler per refresh for the whole session -
    -- invisible, because they all draw the same tooltip.
    go:SetScript("OnClick", function()
        -- Through the REAL slash handler, the same way /valuate verify drives it. Every one
        -- of these commands already prints a considered answer; routing around it to call
        -- the underlying function would put a second, untested printer on screen next to
        -- the tested one.
        local handler = SlashCmdList and SlashCmdList["VALUATE"]
        if row.command and type(handler) == "function" then
            handler((row.command:gsub("^/valuate%s*", "")))
        end
    end)
    go:HookScript("OnEnter", function(self)
        if ns.ShowTooltipSafe(self, "ANCHOR_RIGHT") then
            GameTooltip:AddLine(row.command or "", 1, 1, 1)
            GameTooltip:AddLine("Runs this for you, the same as typing it.", 0.8, 0.8, 0.8, true)
            GameTooltip:Show()
        end
    end)
    go:HookScript("OnLeave", function() GameTooltip:Hide() end)

    rowPool[index] = row
    return row
end

-- The sentence shown when the list is empty.
--
-- This is the most consequential sentence in the addon. An empty to-do list is not read as
-- "no rows"; it is read as "you are done", and people act on it by not acting.
--
-- It used to be fixed text: "Nothing outstanding. Your gear, gems and enchants are all up to
-- date for the scale you are using." Three specific claims, made whether or not any of the
-- three had actually been read. Mid equipment swap the sockets are not read at all, and the
-- panel cheerfully vouched for your gems.
--
-- So the claim now covers exactly what was examined, and no more. Note the wording: "among the
-- things I could check" is narrower than the old sentence and is not an apology - nothing is
-- broken, and a swap finishes on its own. Dressing a transient gap as a fault would send
-- people looking for one.
function ns.TodoEmptyText(unread)
    local all = "Nothing outstanding. Your gear, gems and enchants are all up to date " ..
        "for the scale you are using."
    if type(unread) ~= "table" or #unread == 0 then return all end
    return "Nothing outstanding among the things I could check just now - but " ..
        table.concat(unread, ", ") .. " went unread, so this is not the whole picture. " ..
        "Open this tab again in a moment."
end

-- Said on a NON-empty refresh too, and for the same reason the UI check says it: a list of
-- three jobs is just as much a statement about partial coverage as a clean bill is, and it is
-- the more dangerous of the two to over-read. You do the three things, come back, see nothing,
-- and take that as the window being finished with you.
--
-- Returns nil when everything was read, so the ordinary refresh shows no extra furniture.
function ns.TodoCoverageLine(unread)
    if type(unread) ~= "table" or #unread == 0 then return nil end
    return "Not read this refresh: " .. table.concat(unread, ", ") .. "."
end

function ns.CreateTodoPanel(parent)
    local panel = CreateFrame("Frame", nil, parent)
    panel:SetAllPoints(parent)

    local title = panel:CreateFontString(nil, "OVERLAY", FONT_H2)
    title:SetPoint("TOPLEFT", panel, "TOPLEFT", PADDING, -PADDING)
    title:SetText("What is worth doing")
    title:SetTextColor(unpack(COLORS.textTitle))

    local subtitle = panel:CreateFontString(nil, "OVERLAY", FONT_SMALL)
    subtitle:SetPoint("TOPLEFT", title, "BOTTOMLEFT", 0, -4)
    subtitle:SetPoint("RIGHT", panel, "RIGHT", -PADDING, 0)
    subtitle:SetJustifyH("LEFT")
    subtitle:SetTextColor(unpack(COLORS.textDim))
    subtitle:SetText("In the order that unblocks the rest. Refreshed each time you open this tab.")

    local list = CreateFrame("Frame", nil, panel)
    list:SetPoint("TOPLEFT", subtitle, "BOTTOMLEFT", 0, -ELEMENT_SPACING * 2)
    list:SetPoint("RIGHT", panel, "RIGHT", -PADDING, 0)
    list:SetHeight(1)

    -- Shown when there is nothing to do, which is a real answer and not an error.
    --
    -- BuildTodoList returns an empty list rather than a heading with nothing under it,
    -- deliberately: "a to-do list that always has entries is one you stop opening." The
    -- panel has to say so out loud, or an empty list reads as a panel that failed to load.
    local empty = panel:CreateFontString(nil, "OVERLAY", FONT_BODY)
    empty:SetPoint("TOPLEFT", list, "TOPLEFT", 0, 0)
    empty:SetPoint("RIGHT", panel, "RIGHT", -PADDING, 0)
    empty:SetJustifyH("LEFT")
    empty:SetTextColor(unpack(COLORS.textBody))
    -- Set on every refresh by ns.TodoEmptyText, never here. A default written at build time is
    -- a claim made before anything has been looked at, which is the whole bug.
    empty:SetText("")
    empty:Hide()

    -- What went unread, shown BELOW the rows on a non-empty refresh.
    --
    -- Its own font string rather than a row: "a to-do you cannot act on is not a to-do", and a
    -- swap in flight is not something you can go and do. Putting it on the list would also put
    -- it in the tab's count, which is a number about work.
    local unreadNote = panel:CreateFontString(nil, "OVERLAY", FONT_SMALL)
    unreadNote:SetPoint("TOPLEFT", list, "BOTTOMLEFT", 0, -ELEMENT_SPACING)
    unreadNote:SetPoint("RIGHT", panel, "RIGHT", -PADDING, 0)
    unreadNote:SetJustifyH("LEFT")
    unreadNote:SetTextColor(unpack(COLORS.textDim))
    unreadNote:Hide()

    -- Per-kind accent colours. Keyed by the `kind` field BuildTodoList already sets, so a
    -- new kind appearing there shows up here in the neutral colour rather than erroring -
    -- an unfamiliar entry should still be readable, just not colour-coded yet.
    local KIND_COLOR = {
        scan     = { 1.00, 0.35, 0.35 },  -- nothing below this is known yet; it is not a task, it is a blocker
        scale    = { 1.00, 0.82, 0.00 },  -- your weights are wrong; everything below depends on it
        guess    = { 1.00, 0.55, 0.10 },  -- your weights were never published, only inferred
        upgrade  = { 0.35, 0.90, 0.35 },  -- free, immediate, already in your bags
        bank     = { 0.70, 0.55, 0.95 },  -- yours already, but not from here - needs a trip
        sockets  = { 0.45, 0.70, 1.00 },  -- needs materials
        enchants = { 0.45, 0.70, 1.00 },  -- needs materials
    }
    local NEUTRAL = { 0.6, 0.6, 0.6 }

    local function Refresh()
        -- Two values, and NOT `Valuate.BuildTodoList and Valuate:BuildTodoList()`. Lua adjusts
        -- an `and` expression to a single value, so the second - which is the entire point -
        -- would silently be nil. Written that way first, here and in the to-do list itself,
        -- and caught both times.
        local items, unread = {}, nil
        if Valuate.BuildTodoList then items, unread = Valuate:BuildTodoList() end
        items = items or {}

        for i, item in ipairs(items) do
            local row = rowPool[i] or BuildRow(list, i)
            row.text:SetText(item.text or "")
            if item.detail then
                row.detail:SetText(item.detail)
                row.detail:Show()
            else
                row.detail:SetText("")
                row.detail:Hide()
            end
            ns.SetSolidColor(row.accent, unpack(KIND_COLOR[item.kind] or NEUTRAL))

            -- Data, not a closure. The handlers were installed once when the row was built
            -- and read this field, so a reused row runs the command it is CURRENTLY showing
            -- rather than the one it was first given.
            row.command = item.command

            -- Sized to what is actually in it, not to a number picked when the row was
            -- designed. ROW_HEIGHT is a floor now rather than the height.
            --
            -- These details are sentences, not labels - the one on a guessed scale runs to
            -- two hundred characters and wraps to three lines at this width. A fixed 40px
            -- row put the second and third of them outside their own backdrop, on top of
            -- whatever came next.
            ns.FitRowHeight(row, {
                floor = ROW_HEIGHT, top = TEXT_TOP_PAD, gap = DETAIL_GAP,
                bottom = TEXT_BOTTOM_PAD,
                columns = { { row.text, row.detail } },
            })
            row:Show()
        end

        -- Hide the tail of the pool rather than leaving last visit's entries on screen.
        for i = #items + 1, #rowPool do
            rowPool[i]:Hide()
        end

        if #items == 0 then
            empty:SetText(ns.TodoEmptyText and ns.TodoEmptyText(unread) or "")
            empty:Show()
            list:SetHeight(1)
        else
            empty:Hide()
            -- Summed from the rows themselves, because they no longer share a height.
            local total = 0
            for i = 1, #items do
                total = total + rowPool[i]:GetHeight() + (i > 1 and ROW_GAP or 0)
            end
            list:SetHeight(total)
        end

        -- Only on the non-empty refresh: when the list is empty the sentence above already
        -- carries it, and saying it twice on one screen reads as two different problems.
        local note = ns.TodoCoverageLine and ns.TodoCoverageLine(unread) or nil
        if note and #items > 0 then
            unreadNote:SetText(note)
            unreadNote:Show()
        else
            unreadNote:Hide()
        end

        -- Tell the tab, so the count is on it rather than only inside it.
        --
        -- Guarded because this file loads before ValuateUI.lua, which is where the tab row is
        -- built: the first Refresh runs during the panel's own construction, when there is no
        -- tab to label yet.
        if ns.SetTodoTabCount then ns.SetTodoTabCount(#items) end

        return #items
    end

    Refresh()
    panel.Refresh = Refresh
    ns.RefreshTodoPanel = Refresh
    return panel
end

File: tools/todopanel.js
#!/usr/bin/env node
/*
 * @gate The To Do tab renders the real list, and reuses its rows without mixing them up
 *
 * Builds ui/TodoPanel.lua against a mocked WoW API and a mocked BuildTodoList.
 *
 * tools/todotest.js already owns the hard question - WHAT belongs on the list and in what
 * order - by running the real Valuate:BuildTodoList. This gate deliberately does not repeat
 * any of that. It asks the questions that only exist once the answer is on screen:
 *
 *   - is every item shown, and in the order it was given;
 *   - does an empty list SAY it is empty, rather than showing a blank panel that reads as
 *     one which failed to load;
 *   - are rows reused rather than recreated (WoW frames cannot be destroyed, so a rebuild
 *     per visit leaks one row per item for the whole session);
 *   - and, because they are reused, does each row still run ITS OWN command after the list
 *     underneath it has changed.
 *
 * That last one is the reason this file exists. A handler closing over an item from an
 * earlier refresh keeps working, keeps looking right, and quietly runs the wrong command.
 *
 * Usage:  node tools/todopanel.js
 */
"use strict";

const { load } = require("./luaharness.js");

const run = load([
  "ui/Shared.lua",
  "ui/Animations.lua",
  "ui/Widgets.lua",
  "ui/TodoPanel.lua",
]);

run(
  `
local failures, checks = {}, 0
local function ok(cond, what) checks = checks + 1 if not cond then table.insert(failures, what) end end
local function eq(got, want, what)
    checks = checks + 1
    if got ~= want then
        table.insert(failures, what .. " (got " .. tostring(got) .. ", wanted " .. tostring(want) .. ")")
    end
end

local ns = __ns

GameTooltip = CreateFrame("Frame")
function GameTooltip:SetOwner() end
function GameTooltip:AddLine() end
function GameTooltip:Hide() end
function GameTooltip:Show() end

-- What the slash handler was asked to do, so a click can be checked rather than assumed.
__ran = {}
SlashCmdList = { VALUATE = function(arg) table.insert(__ran, arg) end }

local ITEMS = {}
-- ONE definition, returning BOTH values from the start.
--
-- This began as a one-value mock, and when the second return was added a THIRD-release-later
-- block simply defined the function again further down. Two mocks for one function in one
-- file, with every assertion above the second one still running against the thin copy - the
-- same duplication that had already cost a bug in todotest.js and another in Valuate.lua.
local UNREAD = nil
Valuate.BuildTodoList = function() return ITEMS, UNREAD end

local host = CreateFrame("Frame")
host:SetWidth(600)
host:SetHeight(400)

-- ---- an empty list is an ANSWER ---------------------------------------------------------
-- BuildTodoList returns {} rather than a heading with nothing under it, on purpose: a
-- to-do list that always has entries is one you stop opening. The panel has to say so out
-- loud, or "nothing to do" and "this panel failed to load" look identical.
local panel = ns.CreateTodoPanel(host)
ok(panel ~= nil, "the panel builds")
ok(type(ns.RefreshTodoPanel) == "function", "and publishes a way to refresh it")

local function visibleRows()
    local n = 0
    for _, f in ipairs(__frames) do
        if f.__todoRow and f:IsShown() then n = n + 1 end
    end
    return n
end

local function emptyText()
    for _, f in ipairs(__frames) do
        for _, r in ipairs(f.__regions or {}) do
            if r.GetText and r:GetText() and r:GetText():find("Nothing outstanding", 1, true)
               and r:IsShown() then
                return r:GetText()
            end
        end
    end
    return nil
end

eq(ns.RefreshTodoPanel(), 0, "with nothing to do the list is empty")
ok(emptyText() ~= nil, "and the panel says so, rather than showing a blank")

-- ---- items render in the order they were given -------------------------------------------
ITEMS = {
    { kind = "scale", text = "Refresh Fury - your gear has moved on",
      detail = "Everything below is scored by this scale.", command = "/valuate wizard" },
    { kind = "upgrade", text = "Equip Bonecrusher in Main Hand", command = "/valuate upgrades" },
    { kind = "sockets", text = "Fill 2 empty sockets",
      detail = "Stats you have already earned.", command = "/valuate sockets" },
}
eq(ns.RefreshTodoPanel(), 3, "three things to do gives three rows")
ok(emptyText() == nil, "and the empty message is out of the way")

-- ---- a row with no detail does not leave the last one showing ----------------------------
-- The rows are reused. An item without a detail line landing on a row that had one would
-- keep the previous item's explanation underneath its own text.
local rows = {}
for _, f in ipairs(__frames) do
    if f.__todoRow then rows[#rows + 1] = f end
end
eq(#rows, 3, "exactly three rows were created")
ok(rows[2].detail and not rows[2].detail:IsShown(),
   "an item with no detail hides the detail line rather than inheriting one")
ok(rows[1].detail:IsShown(), "and an item with one shows it")

-- ---- rows grow to fit what is in them -----------------------------------------------------
-- These details are sentences, not labels. The one BuildTodoList writes for a guessed scale
-- runs to two hundred characters and wraps to three lines at this width, and a fixed-height
-- row put the second and third of them outside their own backdrop, over whatever came next.
ITEMS = {
    { kind = "upgrade", text = "Short", command = "/valuate upgrades" },
    { kind = "guess", text = "The weights in Fury are a guess",
      detail = "No stat priority was ever published for that spec, so they were read off its " ..
               "description. Everything below is ranked by them. Edit them as you learn what " ..
               "actually works for you.",
      command = "/valuate scales" },
}
ns.RefreshTodoPanel()
local sized = {}
for _, f in ipairs(__frames) do
    if f.__todoRow and f:IsShown() then sized[#sized + 1] = f end
end
eq(#sized, 2, "two items, two rows")
ok(sized[2]:GetHeight() > sized[1]:GetHeight(),
   "the row with a wrapping paragraph is taller than the one with a single line (" ..
   tostring(sized[1]:GetHeight()) .. " vs " .. tostring(sized[2]:GetHeight()) .. ")")


-- The list frame has to account for it too, or the rows below run off the panel.
local listFrame
for _, f in ipairs(__frames) do
    if f.__todoRow then listFrame = listFrame or f:GetParent() end
end
ok(listFrame and listFrame:GetHeight() >= sized[1]:GetHeight() + sized[2]:GetHeight(),
   "the list is at least as tall as the rows it holds")


-- ---- clicking a row runs THAT row's command ----------------------------------------------
ITEMS = {
    { kind = "scale", text = "Refresh Fury - your gear has moved on",
      detail = "Everything below is scored by this scale.", command = "/valuate wizard" },
    { kind = "upgrade", text = "Equip Bonecrusher in Main Hand", command = "/valuate upgrades" },
    { kind = "sockets", text = "Fill 2 empty sockets",
      detail = "Stats you have already earned.", command = "/valuate sockets" },
}
ns.RefreshTodoPanel()
__ran = {}
rows[1].go.__scripts.OnClick(rows[1].go)
eq(#__ran, 1, "clicking a row runs one command")
eq(__ran[1], "wizard", "and it is that row's own, with the slash prefix stripped")

__ran = {}
rows[3].go.__scripts.OnClick(rows[3].go)
eq(__ran[1], "sockets", "each row runs its own")

-- ---- THE regression the pooling makes possible -------------------------------------------
-- Same frames, different items. A handler that closed over the item from the previous
-- refresh still fires, still looks right, and runs the wrong command entirely.
ITEMS = {
    { kind = "enchants", text = "Enchant 4 items", command = "/valuate enchants" },
}
eq(ns.RefreshTodoPanel(), 1, "the list shrinks to one")
eq(visibleRows(), 1, "and the rows that are no longer needed are hidden, not left on screen")

__ran = {}
rows[1].go.__scripts.OnClick(rows[1].go)
eq(__ran[1], "enchants",
   "the reused row runs the NEW item's command, not the one it was built with")

-- ---- reuse, not recreation ----------------------------------------------------------------
local before = #__frames
ITEMS = {
    { kind = "upgrade", text = "a", command = "/valuate upgrades" },
    { kind = "upgrade", text = "b", command = "/valuate upgrades" },
    { kind = "upgrade", text = "c", command = "/valuate upgrades" },
}
ns.RefreshTodoPanel()
ns.RefreshTodoPanel()
ns.RefreshTodoPanel()
eq(#__frames, before, "refreshing an already-built list creates no new frames")

-- ...and creates no new HANDLERS either, which is the subtler half.
--
-- Anything installed inside the refresh is installed again on every visit to the tab.
-- SetScript merely replaces, so it looks fine; HookScript CHAINS, and the row quietly
-- collects one more tooltip handler per refresh for the rest of the session. Nothing about
-- it is visible - every copy draws the same tooltip - which is exactly why it needs asking.
local tipCalls = 0
local realAdd = GameTooltip.AddLine
GameTooltip.AddLine = function(self, ...) tipCalls = tipCalls + 1 return realAdd(self, ...) end
local probe
for _, f in ipairs(__frames) do
    if f.__todoRow and f:IsShown() then probe = probe or f end
end
if probe then
    probe.go.__scripts.OnEnter(probe.go)
    local once = tipCalls
    for _ = 1, 5 do ns.RefreshTodoPanel() end
    tipCalls = 0
    probe.go.__scripts.OnEnter(probe.go)
    eq(tipCalls, once, "hovering after five refreshes draws the tooltip once, not six times")
end
GameTooltip.AddLine = realAdd

-- ---- an unfamiliar kind is still readable --------------------------------------------------
-- The accent is keyed on the kind BuildTodoList sets. A kind added there and not here must
-- fall back to neutral rather than erroring: an entry nobody colour-coded yet is still an
-- entry you need to be able to read.
ITEMS = { { kind = "somethingNew", text = "a new sort of thing", command = "/valuate report" } }
local fine = pcall(ns.RefreshTodoPanel)
ok(fine, "a kind this panel has never heard of does not break the list")
eq(visibleRows(), 1, "and is still shown")

-- ---- what the list did NOT read, drawn on the panel -----------------------------------------------
--
-- This seam was unexercised. The mock returned ONE value, so unread was always nil, the
-- confident sentence was always the one chosen, and the coverage note added in v0.203.0a had
-- never been rendered by any gate at all.
--
-- The formatters are gated in tools/todounread.js and BuildTodoList's second return is gated in
-- tools/todotest.js. Neither proves the panel joins them up - which is the same shape as the
-- v0.210.0a bug: every part tested, and the path between them not.
--
-- Note the mock now returns TWO values. Lua adjusts an and expression to one, and the panel's
-- own call site carries a comment about exactly that; a fixture that returns one value can
-- never catch it.

-- Every visible string on the panel.
local function panelText()
    local out = {}
    for _, f in ipairs(__frames) do
        if f.IsShown and f:IsShown() and f.GetRegions then
            for _, r in ipairs({ f:GetRegions() }) do
                if r and r.GetText and r:GetText() and r.IsShown and r:IsShown() then
                    out[#out + 1] = r:GetText()
                end
            end
        end
    end
    return table.concat(out, " | ")
end

-- An empty list with nothing unread: the confident sentence, unchanged.
ITEMS, UNREAD = {}, nil
ns.RefreshTodoPanel()
local said = panelText()
ok(said:find("all up to date", 1, true) ~= nil,
   "nothing to do and nothing unread still says your gear is up to date")

-- THE ONE THAT MATTERS. An empty list is read as "you are done", and a source that failed on
-- this refresh means the panel cannot honestly say that.
ITEMS, UNREAD = {}, { "empty sockets (an item is still being swapped)" }
ns.RefreshTodoPanel()
said = panelText()
eq(said:find("all up to date", 1, true), nil,
   "with a source unread, the blanket claim is GONE from the panel - not merely softened")
ok(said:find("sockets", 1, true) ~= nil, "and what was missed is named on screen")

-- Said on a NON-empty refresh too. A list of jobs is just as much a claim about partial
-- coverage as a clean bill is, and the more dangerous of the two to over-read.
ITEMS = { { kind = "scan", text = "Scan your gear", command = "/valuate scan" } }
UNREAD = { "2 worn slots still loading" }
ns.RefreshTodoPanel()
said = panelText()
ok(said:find("Not read this refresh", 1, true) ~= nil,
   "a list with rows on it STILL says what went unread")
ok(said:find("still loading", 1, true) ~= nil, "naming the source")

-- And it withdraws. A note that never clears is one you stop reading.
UNREAD = nil
ns.RefreshTodoPanel()
said = panelText()
eq(said:find("Not read this refresh", 1, true), nil,
   "a clean refresh withdraws the note rather than leaving it up")
ITEMS = {}

return failures, checks
`,
  "todopanel",
  "the To Do tab"
);
